docs/proposals/14-wireguard-based-cluster-vpn.md

```diff
@@ -0,0 +1,178 @@
+
+On the shoot, we create the keys and acquire the WireGuard IP in the standard secret generation. The data is added as a secret to the control plane and to the shootstate. The vpn shoot deployment is extended to include the WireGuard device setup inside the vpn shoot pod network. For certain infrastructures (AWS), we need a re-advertiser to resolve the seed WireGuard endpoint and evaluate whether the IP address changed.
+
+The load balancer exposing the open vpn endpoint is no longer required and therefore removed if this functionality is used.
+
+As we want to slowly expand the usage of the WireGuard solution, we propose to introduce a feature gate for it. Furthermore, since the WireGuard network requires a separate network range, we propose to introduce a new section to the seed settings with two additional flags (enabled & cidr):
+
+```
+apiVersion: core.gardener.cloud/v1beta1
+kind: Seed
+metadata:
+  name: my-seed
+  ...
+spec:
+  ...
+  settings:
+  ...
+    wireguard:
+      enabled: true
+      cidr: 192.168.0.0/22
+```
+
+Last but not least, we propose to introduce an annotation to the shoots to enable/disable the WireGuard tunnel explicitly.
+
+```
+apiVersion: core.gardener.cloud/v1beta1
+kind: Shoot
+metadata:
+  name: my-shoot
+  annotations:
+    alpha.featuregates.shoot.gardener.cloud/wireguard-tunnel: "true"
+  ...
+```
+
+Using this approach, it is easy to switch the solution on and off, i.e. migrate the shoot clusters automatically during ordinary reconciliation.
+
+## High Availability
+
+There is an issue if the node that hosts the WireGuard endpoint fails. The endpoint is migrated to another node however the time required to do this might exceed the budget for downtimes although one could argue that a disruption of less than 30 seconds to 1 minute does not qualify as a downtime and will in almost all cases not noticeable by end users. 
+
+In this case we also assume that TCP connections won't be interrupted - they would just appear to hang. We will confirm this as part of the development and testing effort.
+
+As a mitigation we propose to instantiate 2 Kubelink instances in the seed cluster. The instances must run on different nodes (if possible but we assume a proper seed cluster has more than one node). Each shoot cluster connects to both endpoints. This means that the OpenVPN server is reachable with two different IP addresses. The VPN seed sidecars must attempt to connect to both of them and will continue to do so. The "Persistent Keepalive" feature is set to 21 seconds by default but could be reduced. Due to the redundancy this however appears not to be necessary.
+
+It is desirable that both connections are used in an equal manner. One strategy could be to use the kubelink 1 connection if the first target WireGuard address is even (the last byte of the IPv4 address), otherwise the kubelink 2 connection. The `vpn-seed` sidecars can then use the following configuration in their OpenVPN configuration file:
```

Comment:
Same as above: sidecars are in the seed, they can talk via ClusterIP without going to the LB.

Probably these lines needs correction because I do not get them.

Reply:
Sorry, I don't get the question.

Reply:
Outdated as we are not following this approach.